Update authentication documentation (#4087)
FerretDB/FerretDB: website/docs/security/authentication.md

# Authentication

> FerretDB supports `PLAIN`, `SCRAM-SHA-256` and `SCRAM-SHA-1` authentication mechanisms.

Author: Is this the correct file to update? I see there are versioned documents now and `task docs-dev` runs on versioned one https://github.com/FerretDB/FerretDB/blob/main/website/versioned_docs/version-v1.20/security/authentication.md instead of this document

Reviewer: That's the correct file to edit.

`task docs-dev` renders all files. There is a version selection on the top right.

Author: Great thanks!

> A user with supported mechanism is created by `createUser()` command and stored in the `admin.system` database `users` collection.

Author: @Fashander do we plan to document user management commands?

Reviewer 2: Yes, I'll still add a section on them, and then we can link it.

> A user with supported mechanism is created by `createUser()` command and stored in the `admin.system` database `users` collection.

Reviewer 2: ```suggestion
A user is created using the `createUser()` command and stored in the `admin.system` database `users` collection.
```

Will it still be the same if we remove the "with supported mechanism" since a user is created with the `createUser()` command, with or without the Auth mechanisms?

Author: Yes it will be the same :+1:

> FerretDB uses passed username and password to authenticate against stored credentials.

Reviewer 2: ```suggestion
FerretDB uses the passed username and password to authenticate against the stored credentials.
```

For example, if a client connects as `mongodb://user1:pass1@ferretdb:27018/ferretdb?tls=true&authMechanism=PLAIN`,

> `user1` is authenticated against its credential in `admin.system` database `users` collection.

Reviewer 2: ```suggestion
`user1` is authenticated against its credential stored in `admin.system` database in the `users` collection.
```

> Before the first user is created, the credential passed in the connection string is used to connect directly to the postgreSQL backend.

Reviewer 3: ```suggestion
Before the first user is created in FerretDB, the credential passed in the MongoDB connection string is used to connect directly to the PostgreSQL backend via passthrough.
```

> Before the first user is created, the credential passed in the connection string is used to connect directly to the postgreSQL backend.

Reviewer 2: ```suggestion
Before the first user is created, the credentials provided in the connection string are used to connect directly to the postgreSQL backend.
```

> For example, when `admin.system` database `users` collection is empty and

Reviewer 2: ```suggestion
For instance, when the `admin.system.users` collection is empty and
```

Author: Is that the same meaning as `admin.system` database `users` collection? If that's the case should I replace with that everywhere?

Reviewer 2: I'm assuming the users collection is inside the `admin.system` db. I left the first one, which sets the context for other usages. So you don't necessarily need to do that in all places

Author: I was wrong, it was `admin` database `system.users` collection. Updated that :see_no_evil:

a client connects as `mongodb://pguser1:pgpass1@ferretdb:27018/ferretdb?tls=true&authMechanism=PLAIN`,
it uses `pguser1` to connect to the postgreSQL backend. The `PLAIN` mechanism is used for this case.
Please note this exception no longer applies once the first user is created.

> When usernames and passwords are transferred in plain text,
> the use of [TLS](../security/tls-connections.md) is highly recommended.

Reviewer 3: Maybe

```suggestionx
When usernames and passwords are transferred in plain text or on an unsecured network,
the use of [TLS](../security/tls-connections.md) is highly recommended.
```

# Authentication

> FerretDB supports `PLAIN`, `SCRAM-SHA-256` and `SCRAM-SHA-1` authentication mechanisms.

Reviewer: No, it doesn't yet.

We should document both old and new authentication for now.

Author: Thanks, definitely. Updated.

# Authentication

The authentication is used to verify the identity of the client connecting to FerretDB using username and password.
It blocks the access if the client does not have the valid credentials.

FerretDB provides authentication by relying on the [backend's authentication](#postgresql-backend-authentication) mechanisms by default.
However, there is an [experimental authentication](#experimental-authentication) feature that allows managing and authenticating users within FerretDB.

> Having user management within FerretDB allows creations and deletions of users via command and hides the necessity to have the knowledge of the specific backend to manage users.

Reviewer: "via command" looks a bit strange. That's not a single command. I would say "via normal MongoDB commands" or something

However, FerretDB does not support authorization yet which means fine-grained access control is not yet possible.

## PostgreSQL backend authentication

> For a postgreSQL backend, it uses provided username and password to connect to the postgreSQL backend directly.

Reviewer: PostgreSQL. It is also not clear what "it" is and "postgreSQL backend" is used twice in the same sentence

Author: Updated

docker compose up
```

> A client connects as user `user1` that is authenticated using credentials stored in `admin.system` database `users` collection.

Reviewer 2: ```suggestion
A client connects as the user `user1`, and authenticated using credentials stored in the `admin.system.users` collection.
```

Please refer to your specific driver documentation on how to set this field.

If this does not resolve your issue please file a bug report [here](https://github.com/FerretDB/FerretDB/issues/new?assignees=ferretdb-bot&labels=code%2Fbug%2Cnot+ready&projects=&template=bug.yml).

## Experimental authentication

FerretDB supports experimental `SCRAM-SHA-256` and `SCRAM-SHA-1` authentication mechanisms.
It is enabled by the flag `--test-enable-new-auth` or the environment variable `TestEnableNewAuth`.

> A user is created using the `createUser` command and stored in the `admin` database `system.users` collection.

Reviewer: As discussed, we need a better introduction: what is it, what is the current status, what are upsides and downsides.

Environment variable name is wrong.

FerretDB uses the provided username and password to authenticate against the stored credentials.

> For example, if a client connects as `mongodb://user1:pass1@ferretdb:27018/ferretdb?tls=true&authMechanism=PLAIN`,

Reviewer: There is no reason to use `PLAIN`

Please refer to your specific driver documentation on how to set this field.

If this does not resolve your issue please file a bug report [here](https://github.com/FerretDB/FerretDB/issues/new?assignees=ferretdb-bot&labels=code%2Fbug%2Cnot+ready&projects=&template=bug.yml).

## Experimental authentication

Upon enabling the flag `--test-enable-new-auth` or the environment variable `TEST_ENABLE_NEW_AUTH`,
FerretDB enables user management and authentication against the stored credentials in the `admin` database `system.users` collection.
The users are managed using the `usersInfo`, `createUser`, `updateUser`, `dropUser` and `dropAllUsersFromDatabase` commands.
The users may be created and authenticated using `SCRAM-SHA-256`, `SCRAM-SHA-1` and `PLAIN` mechanisms.

> The `SCRAM` mechanisms are the algorithms used to store hashed passwords. FerretDB also stores hashed password using `PLAIN` mechanism.

Reviewer: If both store hashed passwords, why we highlight that?

The difference is not in the storage, but in the wire protocol exchanges – `PLAIN` requires the client to send the password in the plaintext.

Author: Perhaps we don't need to mention beyond what we support.

> Please note that FerretDB does not support authorization yet, so created users currently have access to all databases.

Reviewer: That's a strange place to say _again_ that authentication is not authorization. Saying it once at the begging is enough, I think.

Author: Removed this one.

> To connect to FerretDB, if a client use `mongodb://user:pass@ferretdb:27018/ferretdb?tls=true&authMechanism=SCRAM-SHA-256`,

Reviewer: "To connect to FerretDB" is not in sync with the reset of the sentence. It could be "to connect to FerretDB, use X" or "for example, if client uses X".

Author: Updated

`user` is authenticated against its credential stored in the `admin.system.users` collection using `SCRAM-SHA-256` authentication mechanism.

> Before the first user is created in FerretDB, the credentials provided in the MongoDB connection string are used to connect directly to the PostgreSQL backend via passthrough.

Reviewer: Let's make it a subsection

Reviewer: Actually, let's make it a subsection later, in #4100. Right now, let's not document it.

Author: removed things related to empty user

For instance, when the `admin.system.users` collection is empty and
a client connects as `mongodb://pguser:pgpass@ferretdb:27018/ferretdb?tls=true&authMechanism=PLAIN`,

> it uses `pguser` to connect to the postgreSQL backend.

Reviewer: I actually don't understand why we rely on `pguser` being in the PostgreSQL. That changes the meaning of the username.

Instead, we should just ignore credentials if `system.users` collection is empty.

Reviewer: Let's not document it now. We will change the behavior and document it in #4100.

Author: :+1: I think https://github.com/FerretDB/FerretDB/issues/4100 would be important before enabling new authentication. Particularly making sure `task mongosh` and `task mongosh-secured` to work after creating users.

When authenticating directly to the backend, only `PLAIN` mechanism is supported.
Please note this exception no longer applies once the first user is created.

> When usernames and passwords are transferred in plain text or on an unsecured network,

Reviewer: Ok, _when_ passwords are transferred in plain text? We don't actually say that.

And usernames are always transferred in plain text and that's fine.

Author: Updated

the use of [TLS](../security/tls-connections.md) is highly recommended.

> ## Example using experimental authentication

Reviewer: It should be either a subsection or not a section at all, not a peer to "old authentication" and "new authentication" sections

Author: Making it a subsection

Start `ferretdb` by specifying `--postgresql-url` with username and password.

> All authenticated clients use `pguser` user to query PostgreSQL backend.

Reviewer: No, we should not do that

Author: removed things related to empty user
